File: be/Media/MediaService.cs
using System.Security.Cryptography;
using FA23_Convocation2023_API.Models;
using Microsoft.EntityFrameworkCore;

namespace FA23_Convocation2023_API.Media;

public sealed class MediaService
{
    private static readonly HashSet<string> OwnerTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "bachelor", "notification", "export", "temp"
    };

    private readonly Convo24Context _db;
    private readonly IObjectStorage _storage;
    private readonly MediaValidator _validator;
    private readonly ILogger<MediaService> _logger;

    public MediaService(Convo24Context db, IObjectStorage storage, MediaValidator validator, ILogger<MediaService> logger)
    {
        _db = db;
        _storage = storage;
        _validator = validator;
        _logger = logger;
    }

    public async Task<MediaAsset> UploadImageAsync(IFormFile file, string ownerType, string ownerId, string actorId, CancellationToken cancellationToken)
    {
        var normalizedOwner = NormalizeOwner(ownerType, ownerId);
        var image = await _validator.ValidateAndNormalizeImageAsync(file, cancellationToken);
        var mediaId = Guid.NewGuid();
        var key = BuildObjectKey(normalizedOwner.Type, normalizedOwner.Id, mediaId);
        var sha = Convert.ToHexString(SHA256.HashData(image.Bytes)).ToLowerInvariant();

        await using var stream = new MemoryStream(image.Bytes, writable: false);
        await _storage.PutAsync(key, stream, stream.Length, image.ContentType, cancellationToken);

        var asset = new MediaAsset
        {
            Id = mediaId,
            ObjectKey = key,
            OriginalName = Path.GetFileName(file.FileName),
            ContentType = image.ContentType,
            Size = image.Bytes.LongLength,
            Width = image.Width,
            Height = image.Height,
            Sha256 = sha,
            OwnerType = normalizedOwner.Type,
            OwnerId = normalizedOwner.Id,
            UploadedBy = actorId
        };

        _db.MediaAssets.Add(asset);
        _db.AuditEvents.Add(new AuditEvent
        {
            Action = "media.upload",
            ActorId = actorId,
            TargetType = nameof(MediaAsset),
            TargetId = asset.Id.ToString(),
            Details = $"owner={asset.OwnerType}:{asset.OwnerId};sha256={asset.Sha256}"
        });

        try
        {
            await _db.SaveChangesAsync(cancellationToken);
        }
        catch
        {
            await _storage.DeleteAsync(key, cancellationToken);
            throw;
        }

        _logger.LogInformation("Media {MediaId} uploaded by {ActorId} for {OwnerType}:{OwnerId}", asset.Id, actorId, asset.OwnerType, asset.OwnerId);
        return asset;
    }

    public Task<MediaAsset?> GetAsync(Guid id, CancellationToken cancellationToken) =>
        _db.MediaAssets.AsNoTracking().FirstOrDefaultAsync(x => x.Id == id && x.Status == MediaAssetStatus.Active, cancellationToken);

    public Task<List<MediaAsset>> ListAsync(CancellationToken cancellationToken) =>
        _db.MediaAssets.AsNoTracking()
            .Where(x => x.Status == MediaAssetStatus.Active)
            .OrderByDescending(x => x.CreatedAt)
            .Take(1000)
            .ToListAsync(cancellationToken);

    public async Task<MediaAsset?> RenameAsync(Guid id, string originalName, string actorId, CancellationToken cancellationToken)
    {
        var asset = await _db.MediaAssets.FirstOrDefaultAsync(x => x.Id == id && x.Status == MediaAssetStatus.Active, cancellationToken);
        if (asset is null) return null;
        var safeName = Path.GetFileName(originalName).Trim();
        if (safeName.Length is < 1 or > 255) throw new MediaValidationException("Invalid display filename.");
        asset.OriginalName = safeName;
        _db.AuditEvents.Add(new AuditEvent { Action = "media.rename", ActorId = actorId, TargetType = nameof(MediaAsset), TargetId = id.ToString() });
        await _db.SaveChangesAsync(cancellationToken);
        return asset;
    }

    public async Task<string?> GetDownloadUrlAsync(Guid id, CancellationToken cancellationToken)
    {
        var asset = await GetAsync(id, cancellationToken);
        if (asset is null) return null;
        await _storage.StatAsync(asset.ObjectKey, cancellationToken);
        return await _storage.PresignGetAsync(asset.ObjectKey);
    }

    public async Task<bool> SoftDeleteAsync(Guid id, string actorId, CancellationToken cancellationToken)
    {
        var asset = await _db.MediaAssets.FirstOrDefaultAsync(x => x.Id == id && x.Status == MediaAssetStatus.Active, cancellationToken);
        if (asset is null) return false;
        asset.Status = MediaAssetStatus.Deleted;
        asset.DeletedAt = DateTimeOffset.UtcNow;
        _db.AuditEvents.Add(new AuditEvent
        {
            Action = "media.delete",
            ActorId = actorId,
            TargetType = nameof(MediaAsset),
            TargetId = id.ToString()
        });
        await _db.SaveChangesAsync(cancellationToken);
        return true;
    }

    public static string BuildObjectKey(string ownerType, string ownerId, Guid mediaId) => ownerType.ToLowerInvariant() switch
    {
        "bachelor" => $"bachelors/{ownerId}/avatar/{mediaId:N}.webp",
        "notification" => $"notifications/{ownerId}/{mediaId:N}.webp",
        "export" => $"exports/{ownerId}/{mediaId:N}.webp",
        _ => $"temp/{ownerId}/{mediaId:N}.webp"
    };

    private static (string Type, string Id) NormalizeOwner(string ownerType, string ownerId)
    {
        var type = ownerType.Trim().ToLowerInvariant();
        if (!OwnerTypes.Contains(type)) throw new MediaValidationException("Unsupported media owner type.");
        if (ownerId.Length is < 1 or > 64 || ownerId.Any(c => !char.IsLetterOrDigit(c) && c != '-'))
            throw new MediaValidationException("Invalid media owner identifier.");
        return (type, ownerId);
    }
}

File: be-nest/src/media/media.service.ts
import { BadRequestException, Inject, Injectable } from '@nestjs/common';
import { createHash, randomUUID } from 'node:crypto';
import { basename } from 'node:path';
import { desc, eq, and } from 'drizzle-orm';
import { DATABASE } from '../database/database.constants.js';
import type { AppDatabase } from '../database/database.types.js';
import { mediaAssets } from '../database/schema/media-schema.js';
import { MediaValidatorService } from './media-validator.service.js';
import { ObjectStorageService } from './object-storage.service.js';

const ownerTypes = new Set(['bachelor', 'notification', 'export', 'temp']);

@Injectable()
export class MediaService {
  constructor(
    @Inject(DATABASE) private readonly database: AppDatabase,
    private readonly storage: ObjectStorageService,
    private readonly validator: MediaValidatorService,
  ) {}

  async upload(file: Express.Multer.File, ownerType: string, ownerId: string, actorId: string) {
    const owner = this.normalizeOwner(ownerType, ownerId);
    const image = await this.validator.validateAndNormalize(file);
    const id = randomUUID();
    const objectKey = this.objectKey(owner.type, owner.id, id);
    const sha256 = createHash('sha256').update(image.bytes).digest('hex');
    await this.storage.put(objectKey, image.bytes, image.contentType);
    try {
      const [asset] = await this.database
        .insert(mediaAssets)
        .values({
          id,
          objectKey,
          originalName: basename(file.originalname),
          contentType: image.contentType,
          size: image.bytes.length,
          width: image.width,
          height: image.height,
          sha256,
          ownerType: owner.type,
          ownerId: owner.id,
          uploadedBy: actorId,
        })
        .returning();
      return asset!;
    } catch (error) {
      await this.storage.delete(objectKey);
      throw error;
    }
  }

  async uploadMany(
    files: Express.Multer.File[],
    ownerType: string,
    ownerId: string,
    actorId: string,
  ) {
    const uploaded: (typeof mediaAssets.$inferSelect)[] = [];
    try {
      for (const file of files) {
        uploaded.push(await this.upload(file, ownerType, ownerId, actorId));
      }
      return uploaded;
    } catch (error) {
      await Promise.allSettled(uploaded.map((asset) => this.rollbackUpload(asset)));
      throw error;
    }
  }

  async get(id: string) {
    const [asset] = await this.database
      .select()
      .from(mediaAssets)
      .where(and(eq(mediaAssets.id, id), eq(mediaAssets.status, 0)))
      .limit(1);
    return asset ?? null;
  }

  list() {
    return this.database
      .select()
      .from(mediaAssets)
      .where(eq(mediaAssets.status, 0))
      .orderBy(desc(mediaAssets.createdAt))
      .limit(1000);
  }

  async rename(id: string, originalName: string) {
    const safeName = basename(originalName).trim();
    if (safeName.length < 1 || safeName.length > 255) {
      throw new BadRequestException({
        code: 'media/invalid-name',
        message: 'Tên hiển thị của tệp không hợp lệ.',
      });
    }
    const [asset] = await this.database
      .update(mediaAssets)
      .set({ originalName: safeName })
      .where(and(eq(mediaAssets.id, id), eq(mediaAssets.status, 0)))
      .returning();
    return asset ?? null;
  }

  async downloadUrl(id: string): Promise<string | null> {
    const asset = await this.get(id);
    if (!asset) return null;
    await this.storage.stat(asset.objectKey);
    return this.storage.presignGet(asset.objectKey);
  }

  async softDelete(id: string): Promise<boolean> {
    const rows = await this.database
      .update(mediaAssets)
      .set({ status: 1, deletedAt: new Date() })
      .where(and(eq(mediaAssets.id, id), eq(mediaAssets.status, 0)))
      .returning();
    return rows.length > 0;
  }

  toDto(asset: typeof mediaAssets.$inferSelect) {
    return {
      id: asset.id,
      originalName: asset.originalName,
      contentType: asset.contentType,
      mimeType: asset.contentType,
      path: `/backend-api/media/${asset.id}/content`,
      size: asset.size,
      width: asset.width,
      height: asset.height,
      ownerType: asset.ownerType,
      ownerId: asset.ownerId,
      createdAt: asset.createdAt,
    };
  }

  objectKey(ownerType: string, ownerId: string, mediaId: string): string {
    const compactId = mediaId.replaceAll('-', '');
    if (ownerType === 'bachelor') return `bachelors/${ownerId}/avatar/${compactId}.webp`;
    if (ownerType === 'notification') return `notifications/${ownerId}/${compactId}.webp`;
    if (ownerType === 'export') return `exports/${ownerId}/${compactId}.webp`;
    return `temp/${ownerId}/${compactId}.webp`;
  }

  private normalizeOwner(ownerType: string, ownerId: string) {
    const type = ownerType.trim().toLowerCase();
    if (!ownerTypes.has(type)) {
      throw new BadRequestException({
        code: 'media/invalid-owner',
        message: 'Loại đối tượng sở hữu tệp không được hỗ trợ.',
      });
    }
    if (!/^[a-zA-Z0-9-]{1,64}$/.test(ownerId)) {
      throw new BadRequestException({
        code: 'media/invalid-owner',
        message: 'Mã đối tượng sở hữu tệp không hợp lệ.',
      });
    }
    return { type, id: ownerId };
  }

  private async rollbackUpload(asset: typeof mediaAssets.$inferSelect): Promise<void> {
    await Promise.allSettled([
      this.storage.delete(asset.objectKey),
      this.database.delete(mediaAssets).where(eq(mediaAssets.id, asset.id)),
    ]);
  }
}
